tasks.py

```diff
@@ -250,10 +258,48 @@ def dev_env(ctx, architecture="amd64", name="kind", cni=None, protocol=None):
     if protocol == "bgp":
         print("Configuring MetalLB with a BGP test environment")
         bgp_dev_env()
+    elif protocol == "layer2":
+        print("Configuring MetalLB with a layer 2 test environment")
+        layer2_dev_env()
     else:
         print("Leaving MetalLB unconfigured")
 
 
+# Configure MetalLB in the dev-env for layer2 testing.
+# Identify the unused network address range from kind network and used it in configmap.
+def layer2_dev_env():
+    # Get unused v4/v6 range for services
+    network_subnet_list = run('docker network inspect -f "{{ '
+            'range .IPAM.Config}}{{.Subnet}} {{end}}" kind', echo=True).stdout.strip().split(' ')
+    used_ipv4_list = run('docker network inspect -f '
+            '"{{range .Containers}}{{.IPv4Address}} {{end}}" kind', echo=True).stdout.strip().split(' ')
+    used_ipv6_list = run('docker network inspect -f '
+            '"{{range .Containers}}{{.IPv6Address}} {{end}}" kind', echo=True).stdout.strip().split(' ')
+    service_range = {}
+    for i in network_subnet_list:
+        network = ipaddress.ip_network(i)
+        if network.version == 6:
+            used_list = sorted(used_ipv6_list)
+        else:
+            used_list = sorted(used_ipv4_list)
+        service_ip_range_start = ipaddress.ip_interface(used_list[-1]) + 1
+        service_ip_range_end = ipaddress.ip_interface(used_list[-1]) + 11
+        if service_ip_range_start not in network:
+            raise Exit(message='network range %s is not in %s' % (service_ip_range_start, network))
+        if service_ip_range_end not in network:
+            raise Exit(message='network range %s is not in %s' % (service_ip_range_end, network))
```

Comment:
Rather than "raise", you could try again to grab a range of 9 and, if that failed, you could try 8 and so on. Otherwise, the developer would need to come into the code, figure out what was going on, and then manually reduce the number.

Reply:
I'm wondering how many service IPs are the minimum. This is a bit dilemma. If developer uses dev-env with assumption that dev-env reserves 10IPs for service, then 9 IPs,  8 IPs or less are not sufficient.

This is just a dev-env script, for developing and testing, so I suppose 10 IPs or die could be simple and reasonable tradeoff.

For example, currently dev-env does similar stuff (i.e. call raise() and quit) if kind cannot the 3nodes.

https://github.com/metallb/metallb/blob/main/tasks.py#L272-L273

Reply:
Probably best not to over-engineer a test script as you suggest. Maybe change the error to state that you are unable add 10 services IPs?